File: public/dsa/two_pointers/sorted_arrays_intersection/optimal/sorted_arrays_intersection.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> intersection(vector<int> &A, vector<int> &B) {
    vector<int> result;
    int i = 0, j = 0;

    while (i < A.size() && j < B.size()) {
        if (A[i] < B[j]) {
            i++;
        } else if (B[j] < A[i]) {
            j++;
        } else {
            result.push_back(A[i]);
            i++;
            j++;
        }
    }

    return result;
}

// tests-start

int main() {
    vector<int> A = {1, 3, 4, 5, 5, 6, 6, 7};
    vector<int> B = {2, 5, 6, 6, 7, 8};
    auto result = intersection(A, B);
    for (int v : result) {
        cout << v << " ";
    }
    cout << endl;
    return 0;
}

// tests-end

/*output
5 6 6 7
*/

File: public/dsa/two_pointers/sorted_arrays_intersection/optimal/sorted_arrays_intersection.py
from typing import List


class Solution:
    def intersection(self, A: List[int], B: List[int]) -> List[int]:
        result = []
        i = 0
        j = 0
        while i < len(A) and j < len(B):
            if A[i] < B[j]:
                i += 1
            elif B[j] < A[i]:
                j += 1
            else:
                result.append(A[i])
                i += 1
                j += 1
        return result


# tests-start
A = [1, 3, 4, 5, 5, 6, 6, 7]
B = [2, 5, 6, 6, 7, 8]
sol = Solution()
print(sol.intersection(A, B))
# tests-end

'''output
[5, 6, 6, 7]
'''
